// calcium.storage/amazon/Connection.ts
module calcium.storage.amazon
{
    export class Region
    {
        static US_EAST_1      = "us-east-1"; // default
        static US_WEST_1      = "us-west-1";
        static EU_WEST_1      = "EU";
        static AP_SOUTHEAST_1 = "ap-southeast-1";
        static AP_SOUTHEAST_2 = "ap-southeast-2";
        static AP_NORTHEAST_1 = "ap-northeast-1";
        static SA_EAST_1      = "sa-east-1";  
    }

    export class Connection
    {
        public accesskey  : string;
        public secretkey  : string;
        public region     : string;
        public bucketname : string;
        public basepath   : string;

        constructor(accesskey : string,
                    secretkey : string,
                    region    : string,
                    bucketname: string,
                    basepath  : string)
        {
            this.accesskey  = accesskey;
            this.secretkey  = secretkey;
            this.region     = region;
            this.bucketname = bucketname;
            this.basepath   = basepath;
        }
    }
}
