// calcium.editor/index.js
var calcium;
(function (calcium) {
    calcium.web = require('calcium.web');
    calcium.template = require('calcium.template');
    calcium.logging = require('calcium.logging');
    calcium.storage = require('calcium.storage');
})(calcium || (calcium = {}));
var __extends = this.__extends || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    __.prototype = b.prototype;
    d.prototype = new __();
};
var calcium;
(function (calcium) {
    (function (editor) {
        (function (routers) {
            var ViewRouter = (function (_super) {
                __extends(ViewRouter, _super);
                function ViewRouter(storage) {
                    var _this = this;
                    _super.call(this);
                    this.storage = storage;

                    this.get('/', function (app) {
                        _this.index(app);
                    });
                }
                ViewRouter.prototype.index = function (context) {
                    calcium.template.storage = this.storage;

                    calcium.template.render("/index.html", null, function (content) {
                        context.response.writeHead(200, { 'content-type': 'text/html' });

                        context.response.end(content);
                    });
                };
                return ViewRouter;
            })(calcium.web.Router);
            routers.ViewRouter = ViewRouter;
        })(editor.routers || (editor.routers = {}));
        var routers = editor.routers;
    })(calcium.editor || (calcium.editor = {}));
    var editor = calcium.editor;
})(calcium || (calcium = {}));
var __extends = this.__extends || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    __.prototype = b.prototype;
    d.prototype = new __();
};
var calcium;
(function (calcium) {
    (function (editor) {
        (function (routers) {
            var StaticRouter = (function (_super) {
                __extends(StaticRouter, _super);
                function StaticRouter(path, storage) {
                    var _this = this;
                    _super.call(this);
                    this.path = path;
                    this.storage = storage;

                    this.get(this.path + "/[path]", function (app) {
                        _this.index(app);
                    });
                }
                StaticRouter.prototype.index = function (context) {
                    var path = calcium.storage.util.path.join(['/', context.parameter.path]);

                    if (calcium.storage.util.path.isfile(path)) {
                        this.storage.readstream(path, function (readstream) {
                            if (!readstream) {
                                context.response.writeHead(404, { 'content-type': 'text/plain' });

                                context.response.end('not found');

                                return;
                            }

                            var mime = calcium.storage.util.mime.fromfilename(path);

                            context.response.writeHead(200, { 'content-type': mime });

                            readstream.pipe(context.response);
                        });

                        return;
                    } else {
                        this.storage.readfolder(path, function (objects) {
                            var json = JSON.stringify(objects, null, ' ');

                            context.response.writeHead(200, { 'content-type': 'application/json' });

                            context.response.end(json);
                        });
                    }
                };
                return StaticRouter;
            })(calcium.web.Router);
            routers.StaticRouter = StaticRouter;
        })(editor.routers || (editor.routers = {}));
        var routers = editor.routers;
    })(calcium.editor || (calcium.editor = {}));
    var editor = calcium.editor;
})(calcium || (calcium = {}));
var __extends = this.__extends || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    __.prototype = b.prototype;
    d.prototype = new __();
};
var calcium;
(function (calcium) {
    (function (editor) {
        (function (routers) {
            var StorageRouter = (function (_super) {
                __extends(StorageRouter, _super);
                function StorageRouter(path, storage) {
                    var _this = this;
                    _super.call(this);
                    this.path = path;
                    this.storage = storage;

                    this.get(this.path, function (context) {
                        _this.index(context);
                    });

                    this.get(this.path + "/[path]", function (context) {
                        _this.index(context);
                    });
                }
                StorageRouter.prototype.index = function (context) {
                    var path = calcium.storage.util.path.join(['/', context.parameter.path ? context.parameter.path : '/']);

                    if (calcium.storage.util.path.isfile(path)) {
                        this.storage.readstream(path, function (readstream) {
                            if (!readstream) {
                                context.response.writeHead(404, { 'content-type': 'text/plain' });

                                context.response.end('not found');

                                return;
                            }

                            var mime = calcium.storage.util.mime.fromfilename(path);

                            context.response.writeHead(200, { 'content-type': mime });

                            readstream.pipe(context.response);
                        });

                        return;
                    } else {
                        this.storage.readfolder(path, function (objects) {
                            var json = JSON.stringify(objects, null, ' ');

                            context.response.writeHead(200, { 'content-type': 'application/json' });

                            context.response.end(json);
                        });
                    }
                };
                return StorageRouter;
            })(calcium.web.Router);
            routers.StorageRouter = StorageRouter;
        })(editor.routers || (editor.routers = {}));
        var routers = editor.routers;
    })(calcium.editor || (calcium.editor = {}));
    var editor = calcium.editor;
})(calcium || (calcium = {}));
var __extends = this.__extends || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    __.prototype = b.prototype;
    d.prototype = new __();
};
var calcium;
(function (calcium) {
    (function (editor) {
        var App = (function (_super) {
            __extends(App, _super);
            function App(directory) {
                _super.call(this);

                this.hub = new calcium.web.Hub();

                this.managed_directory = new calcium.storage.local.Storage({ basepath: directory });

                this.static_directory = new calcium.storage.local.Storage({ basepath: __dirname + "/static" });

                this.view_directory = new calcium.storage.local.Storage({ basepath: __dirname + "/views" });

                this.routers.push(new calcium.editor.routers.StorageRouter('/storage', this.managed_directory));

                this.routers.push(new calcium.editor.routers.ViewRouter(this.view_directory));

                this.routers.push(new calcium.editor.routers.StaticRouter('/static', this.static_directory));
            }
            return App;
        })(calcium.web.App);
        editor.App = App;
    })(calcium.editor || (calcium.editor = {}));
    var editor = calcium.editor;
})(calcium || (calcium = {}));
exports = calcium.editor;
module.exports = exports;
